README.md
#	dafo
__General Date Format__

[![total downloads of dafo](https://img.shields.io/npm/dt/dafo.svg)](https://www.npmjs.com/package/dafo)
[![dafo's License](https://img.shields.io/npm/l/dafo.svg)](https://www.npmjs.com/package/dafo)
[![latest version of dafo](https://img.shields.io/npm/v/dafo.svg)](https://www.npmjs.com/package/dafo)
[![build status of github.com/YounGoat/ecmascript.dafo](https://travis-ci.org/YounGoat/ecmascript.dafo.svg?branch=master)](https://travis-ci.org/YounGoat/ecmascript.dafo)

##	Description

To format date in [PHP](#dafophp) or [MySQL](#dafomysql) style.

##	Table of Contents

*	[DISCLAIMER](#disclaimer)
*	[Get Started](#get-started)
*	[API](#api)
	*	[dafo/mysql](#dafomysql)
	*	[dafo/php](#dafophp)
	*	[Others](#others)
*	[Why dafo](#why-dafo)
*	[About](#about)
*	[References](#references)
*	[CHANGE LOG](./CHANGELOG.md)
*	[Homepage](https://github.com/YounGoat/ecmascript.dafo)

##	DISCLAIMER

I understand that date and datetime is really important and sensitive. However, there are too many details to be dealt with in this package. So, there is __NO guarantee__ that the outputs of __dafo__ will conform to what described in this document.

If anything unexpected found, please create an issue to let me and others know.

##	Get Started

__dafo__ offers different ways to format a date. And you may choose the style which you are familiar with.

###	MySQL Style

```javascript
const DATE_FORMAT = require('dafo/mysql');

// Use customised format mask.
DATE_FORMAT(new Date, '%Y-%M-%d');
// e.g. RETURN 2018-1-1

// Use predefined format mask.
DATE_FORMAT(new Date, DATE_FORMAT.USA);
```

See MySQL Reference Manual [Date and Time Functions, DATE_FORMAT](https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_date-format) for details, and [GET_FORMAT](https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_get-format) for available predefined formats.

###	PHP Style

```javascript
const date_format = require('dafo/php');

date_format(new Date, 'Y-n-j');
// e.g. RETURN 2018-1-1
```

See PHP Manual > [Date/Time Functions](http://php.net/manual/function.date.php) for details.

##	API

###	dafo/mysql

```javascript
const date_format = require('dafo/mysql');
```
*	string __date_format__(Date *date*, string *format*)
*	const string __date_format.DATE.*__
*	const string __date_format.DATETIME.*__
*	const string __date_format.TIME.*__

Special characters prefixed with `%` will be recognized and transformed into corresponding date text. Characters prefixed with `%` but unrecognizable will be preserved wite the leading `%` trimmed.

Follow the next table for frequently used format characters which supported by __dafo/mysql__:

| Catergory  | Example           | Ch     | Meaning                                     |
| :--------- | :---------------- | :----- | :------------------------------------------ |
| Day        | Sun..Sat          | __%a__ | weekday short name |
| Day        | 1st,2nd..31st     | __%D__ | day of the month (ordinal number) |
| Day        | 01..31            | __%d__ | day of the month |
| Day        | 1..31             | __%e__ | day of the month |
| Day        | 001..366          | __%j__ | day of the year |
| Day        | Sunday..Saturday  | __%W__ | weekday name |
| Day        | 0..6              | __%w__ | weekday number (0=Sunday..6=Saturday) |
| Week       | 00..53            | __%U__ | week number of year (mode 0) |
| Week       | 00..53            | __%u__ | week number of year (mode 1) |
| Week       | 01..53            | __%V__ | week number of year (mode 2) |
| Week       | 01..53            | __%v__ | week number of year (mode 3) |
| Month      | Jan..Dec          | __%b__ | month short name |
| Month      | 1..12             | __%c__ | month number |
| Month      | January..December | __%M__ | month name |
| Month      | 01..12            | __%m__ | month number |
| Year       | 1999              | __%X__ | week-numbering year (mode 2) |
| Year       | 1999              | __%x__ | week-numbering year (mode 3) |
| Year       | 2000              | __%Y__ | year |
| Year       | 00                | __%y__ | year |
| Time       | 000000..999999    | __%f__ | micro-seconds |
| Time       | 00..23            | __%H__ | hours (24-hour) |
| Time       | 01..12            | __%h__ | hours (12-hour) |
| Time       | 01..12            | __%I__ | hours (12-hour) |
| Time       | 00..59            | __%i__ | minutes |
| Time       | 0..23             | __%k__ | hours (24-hour) |
| Time       | 1..12             | __%l__ | hours (12-hour) |
| Time       | AM, PM            | __%p__ | Ante meridiem / Post meridiem |
| Time       | 11:59:59 AM       | __%r__ | time (12-hour) |
| Time       | 00..59            | __%S__ | seconds |
| Time       | 00..59            | __%s__ | seconds |
| Time       | 23:59:59          | __%T__ | time (24-hour) |

ATTENTION: Most, but __NOT ALL__ format characters used in MySQL function `DATE_FORMAT()` are supported by __dafo/mysql__.

###	dafo/php

```javascript
const date_format = require('dafo/php');
```
*	string __date_format__(Date *date*, string *format*)

As what `date()` in PHP does, __dafo/php__ will recognize special single characters in format string and transform them into corresponding date text. Unlike __dafo/mysql__, there are no leading escapers before these special characters and those not recognized will be preserved. 

__Escape Character__  
If you want to keep the format character from being replaced, prefix it with an escape character anti-slash `\`, e.g. 
```javascript
date_format(new Date(2000, 0, 1), '\\Y Y');
// RETURN 'Y 2000'

// Because '\' is also the default escaper in JavaScript string, 
// you should input double anti-slash characters '\\' in string literal 
// to create a real '\' in the format.
```

Follow the next table for frequently used format characters which supported by __dafo/php__:

| Catergory  | Example           | Ch    | Meaning                                     |
| :--------- | :---------------- | :---- | :------------------------------------------ |
| Day        | 01..31            | __d__ | day of the month |
| Day        | Mon..Sun          | __D__ | weekday short name |
| Day        | 1..31             | __j__ | day of the month |
| Day        | Sunday..Saturday  | __l__ | weekday name |
| Day        | 1..7              | __N__ | weekday number |
| Day        | st,nd..st         | __S__ | day of the month (ordinal suffix) |1
| Day        | 0..6              | __w__ | weekday number |
| Day        | 0..365            | __z__ | day of the year |
| Week       | 01..53            | __W__ | week number of year |
| Month      | January..December | __F__ | month name |
| Month      | 01..12            | __m__ | month number |
| Month      | Jan..Dec          | __M__ | month short name |
| Month      | 1..12             | __n__ | month number |
| Month      | 28..31            | __t__ | days in the month |
| Year       | 0,1               | __L__ | if it is a leap year |
| Year       | 1999, 2000        | __o__ | week-numbering year |
| Year       | 0100, 1899, 2050  | __Y__ | year A.D. |
| Year       | 00, 99, 50        | __y__ | year A.D. |
| Time       | am, pm            | __a__ | Ante meridiem / Post meridiem |
| Time       | AM, PM            | __A__ | Ante meridiem / Post meridiem |
| Time       | 1..12             | __g__ | 12 hour |
| Time       | 0..23             | __G__ | 24 hour |
| Time       | 01..12            | __h__ | 12 hour |
| Time       | 00..23            | __H__ | 24 hour |
| Time       | 00..59            | __i__ | minutes |
| Time       | 00..59            | __s__ | seconds |
| Time       | 000000..999999    | __u__ | micro-seconds |
| Time       | 000...999         | __v__ | milli-seconds |
| TimeZone   | +0800             | __O__ | offset from GMT (in hours and minutes) |
| TimeZone   | +08:00            | __P__ | offset from GMT (in hours and minutes) |
| TimeZone   | 28800             | __Z__ | offset from GMT (in seconds) |

ATTENTION: Most, but __NOT ALL__ format characters used in PHP function `date()` are supported by __dafo/php__.

###	Others

*	number __dafo.getDayOfYear__(Date *date*)  
	Get the day of the current year.  
	Jan 1st is always 1, and Dec 31st may be 365 (not leap year) or 366 (leap year).

*	number __dafo.getMonthDays__(Date *date*)  
	Get days of the current month.  

*	number __dafo.getWeekOfYear__(Date *date*, number *mode*)  
	Get the week number of year.  
	Read section [Week Mode](#week-mode) for details about the *mode* argument.

*	{ Date *first*, Date *last* } __dafo.getWeekRange__({ number *year*, number *week*, number *mode* })  
	Get the week range.  
	Read section [Week Mode](#week-mode) for details about the *mode* argument.
	
*	{ *year*, *week* } __dafo.getYearWeek__(Date *date*, number *mode*)  
	Get the week number of year, and the corresponding year.  
	In some *mode*, the first few days in a year may be belonging to the last week of last year.   
	Read section [Week Mode](#week-mode) for details about the *mode* argument.

*	boolean __dafo.isLeapYear__(Date *date*)  
	Return `true` if the year of *date* is leap. Otherwise return `false`.

###	Week Mode

The definition of *mode* argument in functions `dafo.getWeekOfYear()` and `dafo.getYearWeek()` is borrowed from [MySQL function `WEEK()`](https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_week). The next table is cited from MySQL Manual and explains the difference between modes: 

| Mode  | 1st weekday | Range | Week 1 is the first week ...  | Follows |
| :---  | :---------- | :---- | :---------------------------- | :------ |
| __0__ | Sunday      | 0-53  | with a Sunday in this year    |
| __1__ | Monday      | 0-53  | with 4 or more days this year |
| __2__ | Sunday      | 1-53  | with a Sunday in this year    |
| __3__ | Monday      | 1-53  | with 4 or more days this year | ISO 8601 |
| __4__ | Sunday      | 0-53  | with 4 or more days this year |
| __5__ | Monday      | 0-53  | with a Monday in this year    |
| __6__ | Sunday      | 1-53  | with 4 or more days this year |
| __7__ | Monday      | 1-53  | with a Monday in this year    |


##  Why *dafo*

There have been lots of packages helping to format date or datetime. Generally, a format string is required to indicate what kind of date string you want, e.g. *YY* in __moment__ means *2 digit year*. The formats used in those packages are similiar, but more or less different from each other. It is really confusing! For those who are familiar with syntax used in `date()` in PHP or `DATE_FORMAT()` in MySQL, it is wasting time to learn another one. 

I am tired of inventing something new but useless. What uou will not meet with in __dafo__ is NOT a new *general* format, but something you have been familiar with.

##  About

##  References

*	MySQL 5.7 Reference Manual > [Date and Time Functions](https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_date-format)
*	PHP Manual > [Date/Time Functions](http://php.net/manual/en/function.date.php)
*	[Date and time format - ISO 8601](https://www.iso.org/iso-8601-date-and-time-format.html) or WIKIPEDIA: [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601)	

CHANGELOG.md
#   dafo Change Log

Notable changes to this project will be documented in this file. This project adheres to [Semantic Versioning 2.0.0](http://semver.org/).

##  [0.1.0] - May 12th, 2019

*   Utility function `dafo.parse()` added.

##	[0.0.1] - 2018-1-17

Released.

---
This CHANGELOG.md follows [*Keep a CHANGELOG*](http://keepachangelog.com/).


test/php.js
'use strict';

const MODULE_REQUIRE = 1
    /* built-in */
    , assert = require('assert')

	/* NPM */
	, noda = require('noda')
    
	/* in-package */
	, date_format = noda.inRequire('php')
	
    ;

describe('php ', () => {
	// 依当时时间创建日期对象，GMT 字段取决于宿主机的时区设定而可能与注释中不同。
	
	let d1 = new Date(2000, 0, 1, 13, 14, 6, 52);
	// Sat Jan 01 2000 13:14:06 GMT+0800 (CST)

	let d2 = new Date(2000, 10, 1, 0, 0, 0, 0);
	// Wed Nov 01 2000 00:00:00 GMT+0800 (CST)

	let d3 = new Date('0990-10-31');
	
	it('d, Day of the month, 2 digits with leading zeros, 01 to 31', () => {
		assert.strictEqual(date_format(d1, 'd'), '01');
	});

	it('D, A textual representation of a day, three letters, Mon through Sun', () => {
		assert.strictEqual(date_format(d1, 'D'), 'Sat');
	});

	it('j, Day of the month without leading zeros, 1 to 31', () => {
		assert.strictEqual(date_format(d1, 'j'), '1');
	});
	
	it('l, A full textual representation of the day of the week, Sunday through Saturday', () => {
		assert.strictEqual(date_format(d1, 'l'), 'Saturday');
	});
	
	it('N, ISO-8601 numeric representation of the day of the week (added in PHP 5.1.0)', () => {
		assert.strictEqual(date_format(d1, 'N'), '6');
		assert.strictEqual(date_format(new Date(2000, 0, 2), 'N'), '7');
	});

	it('S, English ordinal suffix for the day of the month, 2 characters, st, nd, rd or th. Works well with j', () => {
		assert.strictEqual(date_format(d1, 'S'), 'st');
	});

	it('w, Numeric representation of the day of the week, 0 (for Sunday) through 6 (for Saturday)', () => {
		assert.strictEqual(date_format(d1, 'w'), '6');
	});

	it('z, The day of the year (starting from 0)	0 through 365', () => {
		assert.strictEqual(date_format(d1, 'z'), '0');
	});

	it('W, ISO-8601 week number of year, weeks starting on Monday, Example: 42 (the 42nd week in the year)', () => {
		assert.strictEqual(date_format(new Date(2018,  5, 14), 'W'), '24');
		assert.strictEqual(date_format(new Date(2018,  0,  8), 'W'), '02');
		assert.strictEqual(date_format(new Date(2018,  0,  7), 'W'), '01');
		assert.strictEqual(date_format(new Date(2018,  0,  1), 'W'), '01');
		assert.strictEqual(date_format(new Date(2017,  0,  1), 'W'), '52');
	});
	
	it('F, A full textual representation of a month, such as January or March, January through December', () => {
		assert.strictEqual(date_format(d1, 'F'), 'January');
	});

	it('m, Numeric representation of a month, with leading zeros, 01 through 12', () => {
		assert.strictEqual(date_format(d1, 'm'), '01');
	});

	it('M, A short textual representation of a month, three letters, Jan through Dec', () => {
		assert.strictEqual(date_format(d1, 'M'), 'Jan');
	});

	it('n, Numeric representation of a month, without leading zeros, 1 through 12', () => {
		assert.strictEqual(date_format(d1, 'n'), '1');
	});

	it('t, Number of days in the given month, 28 through 31', () => {
		assert.strictEqual(date_format(d1, 't'), '31');
	});

	it('L, Whether it\'s a leap year, 1 if it is a leap year, 0 otherwise.', () => {
		assert.strictEqual(date_format(new Date(2004, 9, 1), 'L'), '1');
		assert.strictEqual(date_format(new Date(2003, 9, 1), 'L'), '0');
		assert.strictEqual(date_format(new Date(2000, 9, 1), 'L'), '1');
		assert.strictEqual(date_format(new Date(1900, 9, 1), 'L'), '0');
	});

	it('o, ISO-8601 week-numbering year', () => {
		assert.strictEqual(date_format(new Date(2018,  5, 14), 'o'), '2018');
		assert.strictEqual(date_format(new Date(2018,  0,  8), 'o'), '2018');
		assert.strictEqual(date_format(new Date(2018,  0,  7), 'o'), '2018');
		assert.strictEqual(date_format(new Date(2018,  0,  1), 'o'), '2018');
		assert.strictEqual(date_format(new Date(2017,  0,  1), 'o'), '2016');
		assert.strictEqual(date_format(d1, 'o'), '1999');
	});

	it('Y, A full numeric representation of a year, 4 digits, Examples: 1999 or 2003', () => {
		assert.strictEqual(date_format(d1, 'Y'), '2000');
	});

	it('y, A two digit representation of a year, Examples: 99 or 03', () => {
		assert.strictEqual(date_format(d1, 'y'), '00');
	});

	it('a, Lowercase Ante meridiem and Post meridiem, am or pm', () => {
		assert.strictEqual(date_format(d1, 'a'), 'pm');
		assert.strictEqual(date_format(d2, 'a'), 'am');
	});

	it('A, Uppercase Ante meridiem and Post meridiem, AM or PM', () => {
		assert.strictEqual(date_format(d1, 'A'), 'PM');
		assert.strictEqual(date_format(d2, 'A'), 'AM');
	});

	it.skip('B, Swatch Internet time, 000 through 999', () => {
		// TODO
	});

	it('g, 12-hour format of an hour without leading zeros, 1 through 12', () => {
		assert.strictEqual(date_format(d1, 'g'), '1');
		assert.strictEqual(date_format(d2, 'g'), '12');
	});

	it('G, 24-hour format of an hour without leading zeros, 0 through 23', () => {
		assert.strictEqual(date_format(d1, 'G'), '13');
		assert.strictEqual(date_format(d2, 'G'), '0');
	});

	it('h, 12-hour format of an hour with leading zeros, 01 through 12', () => {
		assert.strictEqual(date_format(d1, 'h'), '01');
		assert.strictEqual(date_format(d2, 'h'), '12');
	});

	it('H, 24-hour format of an hour with leading zeros, 00 through 23', () => {
		assert.strictEqual(date_format(d1, 'H'), '13');
		assert.strictEqual(date_format(d2, 'H'), '00');
	});

	it('i, Minutes with leading zeros, 00 to 59', () => {
		assert.strictEqual(date_format(d1, 'i'), '14');
		assert.strictEqual(date_format(d2, 'i'), '00');
	});

	it('s, Seconds, with leading zeros, 00 through 59', () => {
		assert.strictEqual(date_format(d1, 's'), '06');
		assert.strictEqual(date_format(d2, 's'), '00');
	});

	it('u, Microseconds (added in PHP 5.2.2)', () => {
		assert.strictEqual(date_format(d1, 'u'), '052000');
	});

	it('v, Milliseconds (added in PHP 7.0.0)', () => {
		assert.strictEqual(date_format(d1, 'v'), '052');
	});

	it.skip('e, Timezone identifier (added in PHP 5.1.0)	Examples: UTC, GMT, Atlantic/Azores', () => {
		// TODO
	});

	it.skip('I, Whether or not the date is in daylight saving time, 1 if Daylight Saving Time, 0 otherwise.', () => {
		// TODO
	});

	it.skip('O, Difference to Greenwich time (GMT) in hours, Example: +0200', () => {
		// TODO
	});

	it.skip('P, Difference to Greenwich time (GMT) with colon between hours and minutes (added in PHP 5.1.3)	Example: +02:00', () => {
		// TODO
	});

	it.skip('T, Timezone abbreviation, Examples: EST, MDT ...', () => {
		// TODO
	});

	it.skip('Z, Timezone offset in seconds.', () => {
		// TODO
	});

	it.skip('c, ISO 8601 date (added in PHP 5)	2004-02-12T15:19:21+00:00', () => {
		// TODO
	});

	it.skip('r, » RFC 2822 formatted date, Example: Thu, 21 Dec 2000 16:01:07 +0200', () => {
		// TODO
	});

	it.skip('U, Seconds since the Unix Epoch (January 1 1970 00:00:00 GMT)', () => {
		// TODO
	});
});

test/util.js
'use strict';

const MODULE_REQUIRE = 1
    /* built-in */
    , assert = require('assert')

	/* NPM */
	, noda = require('noda')
    
	/* in-package */
	, dafo = noda.inRequire('index')
	
    ;

describe('dafo', () => {
	// 依当时时间创建日期对象，GMT 字段取决于宿主机的时区设定而可能与注释中不同。

	let d1 = new Date(2000, 0, 1, 13, 14, 6, 52);
	// Sat Jan 01 2000 13:26:09 GMT+0800 (CST)

	let d2 = new Date(2000, 10, 1, 0, 0, 0, 0);
	// Wed Nov 01 2000 00:00:00 GMT+0800 (CST)

	let WEEK_MODE_TEST_UNITS = [
		{ date: new Date(2018,  5, 14), mode: 0, year: 2018, week: 23 },
		{ date: new Date(2018,  0,  2), mode: 0, year: 2018, week:  0 },
		{ date: new Date(2017,  0,  2), mode: 0, year: 2017, week:  1 },

		{ date: new Date(2018,  5, 14), mode: 1, year: 2018, week:  24 },
		{ date: new Date(2018,  0,  2), mode: 1, year: 2018, week:   1 },
		{ date: new Date(2017,  0,  1), mode: 1, year: 2017, week:   0 },
		{ date: new Date(2017,  0,  9), mode: 1, year: 2017, week:   2 },
		{ date: new Date(2017,  0,  8), mode: 1, year: 2017, week:   1 },
		{ date: new Date(2017,  0,  2), mode: 1, year: 2017, week:   1 },

		{ date: new Date(2018,  5, 14), mode: 2, year: 2018, week:  23 },
		{ date: new Date(2018,  0,  7), mode: 2, year: 2018, week:   1 },
		{ date: new Date(2018,  0,  6), mode: 2, year: 2017, week:  53 },
		{ date: new Date(2018,  0,  2), mode: 2, year: 2017, week:  53 },
		{ date: new Date(2017,  0,  2), mode: 2, year: 2017, week:   1 },

		{ date: new Date(2018,  5, 14), mode: 3, year: 2018, week:  24 },
		{ date: new Date(2018,  0,  8), mode: 3, year: 2018, week:   2 },
		{ date: new Date(2018,  0,  7), mode: 3, year: 2018, week:   1 },
		{ date: new Date(2018,  0,  1), mode: 3, year: 2018, week:   1 },
		{ date: new Date(2017,  0,  1), mode: 3, year: 2016, week:  52 },

		{ date: new Date(2018,  5, 14), mode: 4, year: 2018, week:  24 },
		{ date: new Date(2018,  0,  7), mode: 4, year: 2018, week:   2 },
		{ date: new Date(2018,  0,  6), mode: 4, year: 2018, week:   1 },
		{ date: new Date(2018,  0,  1), mode: 4, year: 2018, week:   1 },
		{ date: new Date(2016,  0,  3), mode: 4, year: 2016, week:   1 },
		{ date: new Date(2016,  0,  1), mode: 4, year: 2016, week:   0 },

		{ date: new Date(2018,  5, 14), mode: 5, year: 2018, week:  24 },
		{ date: new Date(2018,  0,  8), mode: 5, year: 2018, week:   2 },
		{ date: new Date(2018,  0,  7), mode: 5, year: 2018, week:   1 },
		{ date: new Date(2018,  0,  1), mode: 5, year: 2018, week:   1 },
		{ date: new Date(2017,  0,  2), mode: 5, year: 2017, week:   1 },
		{ date: new Date(2017,  0,  1), mode: 5, year: 2017, week:   0 },

		{ date: new Date(2018,  5, 14), mode: 6, year: 2018, week:  24 },
		{ date: new Date(2018,  0,  7), mode: 6, year: 2018, week:   2 },
		{ date: new Date(2018,  0,  1), mode: 6, year: 2018, week:   1 },
		{ date: new Date(2016,  0,  3), mode: 6, year: 2016, week:   1 },
		{ date: new Date(2016,  0,  2), mode: 6, year: 2015, week:  52 },
		{ date: new Date(2016,  0,  1), mode: 6, year: 2015, week:  52 },

		{ date: new Date(2018,  5, 14), mode: 7, year: 2018, week:  24 },
		{ date: new Date(2018,  0,  8), mode: 7, year: 2018, week:   2 },
		{ date: new Date(2018,  0,  7), mode: 7, year: 2018, week:   1 },
		{ date: new Date(2018,  0,  1), mode: 7, year: 2018, week:   1 },
		{ date: new Date(2016,  0,  4), mode: 7, year: 2016, week:   1 },
		{ date: new Date(2016,  0,  3), mode: 7, year: 2015, week:  52 },
		{ date: new Date(2016,  0,  1), mode: 7, year: 2015, week:  52 },
	];

	it('getDayOfYear()', () => {
		assert.equal(dafo.getDayOfYear(d1), 1);
		assert.equal(dafo.getDayOfYear(d2), 306);
	});

	it('getMonthDays()', () => {
		assert.equal(dafo.getMonthDays(new Date(2000, 0, 1)), 31);
		assert.equal(dafo.getMonthDays(new Date(2000, 1, 1)), 29);
		assert.equal(dafo.getMonthDays(new Date(2000, 3, 1)), 30);
		assert.equal(dafo.getMonthDays(new Date(2001, 1, 1)), 28);
	});

	it('getWeekOfYear()', () => {
		WEEK_MODE_TEST_UNITS.forEach(unit => {
			let ret = dafo.getWeekOfYear(unit.date, unit.mode);
			assert.equal(ret, unit.week);
		});
	});
	
	it('getYearWeek()', () => {
		WEEK_MODE_TEST_UNITS.forEach(unit => {
			let ret = dafo.getYearWeek(unit.date, unit.mode);
			assert.equal(ret.year, unit.year);
			assert.equal(ret.week, unit.week);
		});
	});

	it('isLeapYear()', () => {
		assert( dafo.isLeapYear(new Date(2000, 0, 1)));
		assert( dafo.isLeapYear(new Date(2004, 0, 1)));
		assert(!dafo.isLeapYear(new Date(1900, 0, 1)));
		assert(!dafo.isLeapYear(new Date(2001, 0, 1)));
	});
	
});

getYearWeek.js

'use strict';

const MODULE_REQUIRE = 1
    /* built-in */
    
    /* NPM */
    
    /* in-package */
    , getDayOfYear = require('./getDayOfYear')
    ;

/**
 * Get the week of the year, 0..53 in default mode 0.
 * @param {Date}     d
 * @param {number}  [mode=0]
 * 
 * -- READMORE --
 * https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_week
 *     The following table describes how the mode argument works.
 *     Mode	First day of week	Range	Week 1 is the first week …
 *     0	Sunday	0-53	with a Sunday in this year
 *     1	Monday	0-53	with 4 or more days this year
 *     2	Sunday	1-53	with a Sunday in this year
 *     3	Monday	1-53	with 4 or more days this year
 *     4	Sunday	0-53	with 4 or more days this year
 *     5	Monday	0-53	with a Monday in this year
 *     6	Sunday	1-53	with 4 or more days this year
 *     7	Monday	1-53	with a Monday in this year
 * 
 */
function getYearWeek(d, mode = 0) {
    // Get the weekday of Jan 1st.
    let Jan1st = new Date(d);
    Jan1st.setMonth(0);
    Jan1st.setDate(1);
    let weekdayOfJan1st = Jan1st.getDay();

    // The offset from Jan 1st to the first day of week 1.
    // 0 means Jan 1st is just the first day of week 1.
    // Negative offset means the first day of week 1 is actually in last year.
    let week1Offset = null;
    
    // First day of week: Sunday
    // Week 1 is the first week with a Sunday in this year. Namely, 
    // Week 1 is the first week with a Thursday in this year.
    if (mode == 0 || mode == 2) {
        week1Offset = (7 - weekdayOfJan1st) % 7;
    } 

    // First day of week: Monday
    // Week 1 is the first week with a Monday in this year.
    else if (mode == 5 || mode == 7) {
        week1Offset = (8 - weekdayOfJan1st) % 7;
    }

    // First day of week: Sunday
    // Week 1 is the first week with 4 or more days this year.
    else if (mode == 4 || mode == 6) {
        week1Offset = [0,1,2,3].includes(weekdayOfJan1st) 
            ? 0 - weekdayOfJan1st
            : 7 - weekdayOfJan1st
            ;
    } 

    // First day of week: Monday
    // Week 1 is the first week with 4 or more days this year.
    else if (mode == 1 || mode == 3) {
        week1Offset = [1,2,3,4].includes(weekdayOfJan1st) 
            ? 1 - weekdayOfJan1st
            : (8 - weekdayOfJan1st) % 7
            ;
    }
    
    let year = d.getFullYear();
    let daysOfYear = getDayOfYear(d);

    // In these modes, dates before week 1 belong to the last week of last year.
    if (daysOfYear <= week1Offset && [2,3,6,7].includes(mode)) {
        let Dec31st = new Date(Jan1st);
        Dec31st.setDate(0);
        return getYearWeek(Dec31st, mode);
    }
    else {
        let week = Math.ceil((daysOfYear - week1Offset) /  7);
        if (week == 0) week = 0; // Avoid -0.
        return { year, week };
    }
}

module.exports = getYearWeek;

php.js
/**
 * SEE
 *   PHP Manual > Function Reference > Date and Time Related Extensions > Date/Time > Date/Time Functions
 *   http://php.net/manual/en/function.date.php
 */

'use strict';

const MODULE_REQUIRE = 1
	/* built-in */
	
	/* NPM */
	, noda = require('noda')
	
	/* in-package */
	, EN = noda.inRequire('EN')
	, ordinal = noda.inRequire('lib/ordinal')
	, getDayOfYear = noda.inRequire('getDayOfYear')
	, getMonthDays = noda.inRequire('getMonthDays')
	, getYearWeek = noda.inRequire('getYearWeek')
	, isLeapYear = noda.inRequire('isLeapYear')

	/* in-file */
	, formatI = (n, digits) => '0'.repeat(digits - (n+'').length) + n
	, getGMTOffset = (d) => {
		let offset = d.getTimezoneOffset();
		let sign = offset > 0 ? '-' : '+';
		offset = Math.abs(offset);
		let hours = Math.floor(offset / 60);
		let minutes = (offset % 60);
		return { sign, H: formatI(hours, 2), i: formatI(hours, 2) };
	}
	;

const ESCAPE_CHAR = '\\';

const FORMATTER = {
	// Day	---	---
	// d	Day of the month, 2 digits with leading zeros	01 to 31
	d : d => formatI(d.getDate(), 2),

	// D	A textual representation of a day, three letters	Mon through Sun
	D : d => EN.weekdayShortNames[d.getDay()],

	// j	Day of the month without leading zeros	1 to 31
	j : d => d.getDate(),

	// l (lowercase 'L')	A full textual representation of the day of the week	Sunday through Saturday
	l : d => EN.weekdayNames[d.getDay()],

	// N	ISO-8601 numeric representation of the day of the week (added in PHP 5.1.0)	1 (for Monday) through 7 (for Sunday)
	N : d => {
		let n = d.getDay();
		return (n == 0 ? 7 : n) + '';
	},

	// S	English ordinal suffix for the day of the month, 2 characters	st, nd, rd or th. Works well with j
	S : d => ordinal.suffix(d.getDate()),

	// w	Numeric representation of the day of the week	0 (for Sunday) through 6 (for Saturday)
	w : d => d.getDay() + '',
	
	// z	The day of the year (starting from 0)	0 through 365
	z : d => (getDayOfYear(d) - 1) + '',

	// Week	---	---
	// W	ISO-8601 week number of year, weeks starting on Monday	Example: 42 (the 42nd week in the year)
	W : d => formatI(getYearWeek(d, 3).week, 2),

	// Month	---	---
	// F	A full textual representation of a month, such as January or March	January through December
	F : d => EN.monthNames[d.getMonth()],

	// m	Numeric representation of a month, with leading zeros	01 through 12
	m : d => formatI(d.getMonth() + 1, 2),

	// M	A short textual representation of a month, three letters	Jan through Dec
	M : d => EN.monthShortNames[d.getMonth()],

	// n	Numeric representation of a month, without leading zeros	1 through 12
	n : d => (d.getMonth() + 1) + '',

	// t	Number of days in the given month	28 through 31
	t : d => getMonthDays(d) + '',

	// Year	---	---
	// L	Whether it's a leap year	1 if it is a leap year, 0 otherwise.
	L : d => isLeapYear(d) ? '1' : '0',

	// o	ISO-8601 week-numbering year. This has the same value as Y, except that if the ISO week number (W) belongs to the previous or next year, that year is used instead. (added in PHP 5.1.0)	Examples: 1999 or 2003
	o : d => getYearWeek(d, 3).year + '',

	// Y	A full numeric representation of a year, 4 digits	Examples: 1999 or 2003
	Y : d => formatI(d.getFullYear(), 4),

	// y	A two digit representation of a year	Examples: 99 or 03
	y : d => formatI(d.getFullYear(), 4).slice(-2),

	// Time	---	---
	// a	Lowercase Ante meridiem and Post meridiem	am or pm
	a : d => d.getHours() < 12 ? 'am' : 'pm',

	// A	Uppercase Ante meridiem and Post meridiem	AM or PM
	A : d => d.getHours() < 12 ? 'AM' : 'PM',

	// B	Swatch Internet time	000 through 999
	
	// g	12-hour format of an hour without leading zeros	1 through 12
	g : d => {
		let hours = d.getHours() % 12;
		return (hours == 0 ? 12 : hours) + '';
	},

	// G	24-hour format of an hour without leading zeros	0 through 23
	G : d => d.getHours() + '',

	// h	12-hour format of an hour with leading zeros	01 through 12
	h : d => {
		let hours = d.getHours() % 12;
		return formatI(hours == 0 ? 12 : hours, 2);
	},

	// H	24-hour format of an hour with leading zeros	00 through 23
	H : d => formatI(d.getHours(), 2),

	// i	Minutes with leading zeros	00 to 59
	i : d => formatI(d.getMinutes(), 2),

	// s	Seconds, with leading zeros	00 through 59
	s : d => formatI(d.getSeconds(), 2),

	// u	Microseconds (added in PHP 5.2.2). Note that date() will always generate 000000 since it takes an integer parameter, whereas DateTime::format() does support microseconds if DateTime was created with microseconds.	Example: 654321
	u : d => formatI(d.getMilliseconds(), 3) + '000',

	// v	Milliseconds (added in PHP 7.0.0). Same note applies as for u.	Example: 654
	v : d => formatI(d.getMilliseconds(), 3),

	// Timezone	---	---
	// e	Timezone identifier (added in PHP 5.1.0)	Examples: UTC, GMT, Atlantic/Azores

	// I (capital i)	Whether or not the date is in daylight saving time	1 if Daylight Saving Time, 0 otherwise.

	// O	Difference to Greenwich time (GMT) in hours	Example: +0200
	O : d => {
		let o = getGMTOffset(d);
		return `${o.sign}${o.H}${o.i}`;
	},

	// P	Difference to Greenwich time (GMT) with colon between hours and minutes (added in PHP 5.1.3)	Example: +02:00
	P : d => {
		let o = getGMTOffset(d);
		return `${o.sign}${o.H}:${o.i}`;
	},

	// T	Timezone abbreviation	Examples: EST, MDT ...
	
	// Z	Timezone offset in seconds. The offset for timezones west of UTC is always negative, and for those east of UTC is always positive.	-43200 through 50400
	Z : d => d.getTimezoneOffset() * 60 + '',
	
	// Full Date/Time	---	---
	// c	ISO 8601 date (added in PHP 5)	2004-02-12T15:19:21+00:00
	c : d => date_format(d, 'Y-m-d\\TH:i:sP'),

	// r	» RFC 2822 formatted date	Example: Thu, 21 Dec 2000 16:01:07 +0200
	r : d => date_format(d, 'D, d M Y H:i:s O'),

	// U	Seconds since the Unix Epoch (January 1 1970 00:00:00 GMT)	See also time()
	U : d => parseInt(d.getTime() / 1000) + '',
};

const format = (d, char) => {
	let fn = FORMATTER[char];
	if (typeof fn == 'string') {
		return format(d, fn);
	}
	else if (fn) {
		return fn(d);
	}
	else {
		return char;
	}
};

function date_format(date, mask) {
	let escaped = false, cursor = 0;
	let output = '';
	for (let cursor = 0; cursor < mask.length; cursor++) {
		let char = mask[cursor];
		if (escaped) {
			output += char;
		}
		else if (char == ESCAPE_CHAR) {
			escaped = true;
		}
		else {
			output += format(date, char);
		}
	}
	return output;
}

module.exports = date_format;